components/script/dom/trackevent.rs

```diff
@@ -0,0 +1,137 @@
+use crate::dom::bindings::codegen::Bindings::TrackEventBinding;
+use crate::dom::bindings::codegen::Bindings::TrackEventBinding::TrackEventMethods;
+use crate::dom::bindings::codegen::UnionTypes::VideoTrackOrAudioTrackOrTextTrack;
+use crate::dom::bindings::error::Fallible;
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::refcounted::Trusted;
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject};
+use crate::dom::bindings::root::Dom;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::event::Event;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::texttrack::TextTrack;
+use crate::dom::videotrack::VideoTrack;
+use crate::dom::window::Window;
+use crate::task_source::TaskSource;
+use dom_struct::dom_struct;
+use servo_atoms::Atom;
+
+#[must_root]
+#[derive(JSTraceable, MallocSizeOf)]
+enum MediaTrack {
+    Video(Dom<VideoTrack>),
+    Audio(Dom<AudioTrack>),
+    Text(Dom<TextTrack>),
+}
+
+#[dom_struct]
+pub struct TrackEvent {
+    event: Event,
+    track: Option<MediaTrack>,
+}
+
+impl TrackEvent {
+    #[allow(unrooted_must_root)]
+    fn new_inherited(track: &Option<VideoTrackOrAudioTrackOrTextTrack>) -> TrackEvent {
+        let media_track = match track {
+            Some(VideoTrackOrAudioTrackOrTextTrack::VideoTrack(VideoTrack)) => {
+                Some(MediaTrack::Video(Dom::from_ref(VideoTrack)))
+            },
+            Some(VideoTrackOrAudioTrackOrTextTrack::AudioTrack(AudioTrack)) => {
+                Some(MediaTrack::Audio(Dom::from_ref(AudioTrack)))
+            },
+            Some(VideoTrackOrAudioTrackOrTextTrack::TextTrack(TextTrack)) => {
+                Some(MediaTrack::Text(Dom::from_ref(TextTrack)))
+            },
+            None => None,
+        };
+
+        TrackEvent {
+            event: Event::new_inherited(),
+            track: media_track,
+        }
+    }
+
+    pub fn new(
+        global: &GlobalScope,
+        type_: Atom,
+        bubbles: bool,
+        cancelable: bool,
+        track: &Option<VideoTrackOrAudioTrackOrTextTrack>,
+    ) -> DomRoot<TrackEvent> {
+        let te = reflect_dom_object(
+            Box::new(TrackEvent::new_inherited(&track)),
+            global,
+            TrackEventBinding::Wrap,
+        );
+        {
+            let event = te.upcast::<Event>();
+            event.init_event(type_, bubbles, cancelable);
+        }
+        te
+    }
+
+    pub fn Constructor(
+        window: &Window,
+        type_: DOMString,
+        init: &TrackEventBinding::TrackEventInit,
+    ) -> Fallible<DomRoot<TrackEvent>> {
+        Ok(TrackEvent::new(
+            &window.global(),
+            Atom::from(type_),
+            init.parent.bubbles,
+            init.parent.cancelable,
+            &init.track,
+        ))
+    }
+
+    pub fn queue_track_event(&self) {
```

Comment:
Given that this is only called by one location, I don't think we need a method for this operation.